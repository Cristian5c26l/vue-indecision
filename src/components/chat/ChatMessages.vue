<template>
  <div ref="chatRef" class="flex-1 overflow-y-auto p-4">
    <div class="flex flex-col space-y-2">
      <!-- Messages go here -->
      <!-- <ChatBubble :itsMine="true" :message="'Quieres tomar un café?'" />
      <ChatBubble
        :itsMine="false"
        :message="'no'"
        :image="'https://yesno.wtf/assets/no/22-8806dbccb1edf544723b7f095ff722e8.gif'"
      /> -->

      <!-- <ChatBubble
        v-for="message in messages"
        :key="message.id"
        :its-mine="message.itsMine"
        :message="message.message"
        :image="message.image"
      /> -->

      <!-- :is-last="index === messages.length - 1" -->
      <ChatBubble v-for="message in messages" :key="message.id" v-bind="message" />
      <LoaderDots v-if="isTyping" />
    </div>
  </div>
</template>

<script lang="ts" setup>
import ChatBubble from '@/components/chat/ChatBubble.vue';
import type { ChatMessage } from '@/interfaces/chat-message.interface';
import LoaderDots from './LoaderDots.vue';
import { ref, watch } from 'vue';

interface Props {
  messages: ChatMessage[];
  isTyping: boolean;
}

const props = defineProps<Props>();

const chatRef = ref<HTMLDivElement | null>(null);

watch(props, () => {
  // funcion watch estará observando "messages". Cuando detecte que cambia propiedad messages, se ejecutará la funcion de flecha que hace que se haga el scroll hasta abajo. Es necesario que se ejecute despues de 100 milisegundos para que los mensajes (ChatBubbles) se hayan renderizado por completo para despues hacer el scrollTo
  // console.log('messages cambió');
  setTimeout(() => {
    chatRef.value?.scrollTo({
      top: chatRef.value.scrollHeight, // posicion hacia donde se realizará el scroll
      behavior: 'smooth', // smooth hace que el scroll del elemento chatRef haga un scroll lento
    });
  }, 100);
});
</script>
