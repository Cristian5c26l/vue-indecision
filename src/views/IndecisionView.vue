<!-- Fuente: https://tailwindcomponents.com/component/chat-layout -->
<template>
  <div class="bg-gray-100 h-screen flex flex-col max-w-lg mx-auto">
    <div class="bg-blue-500 p-4 text-white flex items-center gap-4">
      <img
        class="w-12 h-12 object-cover rounded-full"
        src="https://media.istockphoto.com/id/1437816897/photo/business-woman-manager-or-human-resources-portrait-for-career-success-company-we-are-hiring.jpg?s=612x612&w=0&k=20&c=tyLvtzutRh22j9GqSGI33Z4HpIwv9vL_MZw_xOE19NQ%3D"
        alt=""
      />
      <div class="dark:text-white">
        <span>Mi esposa</span
        ><!-- span or div does work! -->
        <div class="text-sm text-gray-500 dark:text-gray-300">En línea</div>
      </div>
    </div>

    <!--ChatMessages-->
    <ChatMessages :messages="messages" :is-typing="isTyping" />

    <!--MessageBox-->
    <MessageBox v-on:send-message="onSendMessage" />
  </div>
</template>

<script lang="ts" setup>
import ChatMessages from '@/components/chat/ChatMessages.vue';
import MessageBox from '@/components/chat/MessageBox.vue';
import { useChat } from '@/composables/useChat';

const { messages, onSendMessage, isTyping } = useChat();

// Exportar los métodos explícitamente para poder acceder a ellos en las pruebas
defineExpose({
  onSendMessage,
});
</script>
