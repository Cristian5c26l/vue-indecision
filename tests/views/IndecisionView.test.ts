import ChatMessages from '@/components/chat/ChatMessages.vue';
import MessageBox from '@/components/chat/MessageBox.vue';
import IndecisionView from '@/views/IndecisionView.vue';
import { mount } from '@vue/test-utils';

const chatMockMessages = {
  template: '<div data-test-id="mock-messages">Mock ChatMessages</div>',
};

// ESTE MOCKEO SI FUNCIONA. mockear el retorno de la funcion useChat contenida en el archivo useChat.ts. Un ejercicio similar esta en https://vitest.dev/api/vi.html#vi-dounmock
// vi.mock('@/composables/useChat', () => ({
//   useChat: () => ({
//     messages: [],
//     isTyping: false,
//     onSendMessage: vi.fn(),
//   }),
// }));

// ESTE MOCKEO FUNCIONA
// Para la prueba should calls onSendMessage when sending a message (version 2).
//ESTA ES UNA PREPARACION PARA QUE CUANDO SE IMPORTE FUNCION useChat que está en useChat.ts y despues se recuperé onSendMessage de dicha funcion, onSendMessage sea un mock function
// vi.mock('@/composables/useChat', async (importOriginal) => {
//   const mod = await importOriginal<typeof import('@/composables/useChat')>(); // mod literalmente contiene todo lo exportado por @/composables/useChat

//   return {
//     ...mod, // esparcir toda funcion, variable u objeto contenido en mod (@/composables/useChat.ts)
//     useChat: vi.fn(() => ({
//       // Funcion useChat, que está dentro de mod, lo redefinimos, de modo que sea una funcion que retorne tambien un objeto con la propiedad onSendMessage la cual sea una funcion mockeada (vi.fn()), pero mantenga las otras propiedades originales contenidas en el objeto devuelto por funcion useChat() la cual es exportada desde @/composables/useChat
//       ...mod.useChat(), // Mantener propiedades devueltas por funcion useChat que está dentro de mod
//       onSendMessage: vi.fn(), // Mockea solo onSendMessage
//     })),
//   };
// });

describe('<IndecisionView />', () => {
  test('should renders chat messages and messagebox correctly', () => {
    const wrapper = mount(IndecisionView);

    expect(wrapper.html()).toMatchSnapshot();

    expect(wrapper.findComponent(ChatMessages).exists()).toBe(true);
    expect(wrapper.findComponent(MessageBox).exists()).toBe(true);
  });

  test('should calls onSendMessage when sending a message', async () => {
    // const consoleLogSpy = vi.spyOn(console, 'log');

    const onSendMessageMock = vi.fn();
    const wrapper = mount(IndecisionView, {
      global: {
        stubs: {
          ChatMessages: chatMockMessages,
        },
        mocks: {
          // mockear la funcion onSendMessage que vive en el script setup de IndecisionView.vue y que se ejecuta cuando MessageBox emite el evento sendMessage mandandole un mensaje a dicha funcion onSendMessage
          onSendMessage: onSendMessageMock,
        },
      },
    });
    // console.log(wrapper.vm); // muestra { onSendMessage: [AsyncFunction: onSendMessage] } solo si desde IndecisionView.vue onSendMessage se expone con defineExpose y cuando no se tiene vi.mock('@/composables/useChat'
    // Cuando se tiene mocks onSendMessage: onSendMessageMock, este ultimo sobreescribe el onSendMessage de useChat original

    // (wrapper.vm as any).onSendMessage = onSendMessageMock; // esto no funciona

    // Simular evento personalizado
    const messageBoxComponent = wrapper.findComponent(MessageBox); // messageBoxComponent es una instancia de MessageBox
    const message = 'Hola Mundo';
    messageBoxComponent.vm.$emit('sendMessage', message); // Esto hace que, al emitir el evento sendMessage (send-message del lado de vue y html) desde MessageBox, se ejecute la funcion asociada a dicho evento. Esta funcion sería onSendMessage la cual va a recibir

    // await new Promise((r) => setTimeout(r, 150));

    expect(onSendMessageMock).toHaveBeenCalled();
    expect(onSendMessageMock).toHaveBeenCalledOnce();
    expect(onSendMessageMock).toHaveBeenCalledTimes(1);
    expect(onSendMessageMock).toHaveBeenCalledWith(message);
  });

  // test('should calls onSendMessage when sending a message (version 2)', async () => {
  //   vi.mock('@/composables/useChat', async (importOriginal) => {
  //     const mod = await importOriginal<typeof import('@/composables/useChat')>(); // mod literalmente contiene todo lo exportado por @/composables/useChat

  //     return {
  //       ...mod, // esparcir toda funcion, variable u objeto contenido en mod (@/composables/useChat.ts)
  //       useChat: vi.fn(() => ({
  //         // Funcion useChat, que está dentro de mod, lo redefinimos, de modo que sea una funcion que retorne tambien un objeto con la propiedad onSendMessage la cual sea una funcion mockeada (vi.fn()), pero mantenga las otras propiedades originales contenidas en el objeto devuelto por funcion useChat() la cual es exportada desde @/composables/useChat
  //         ...mod.useChat(), // Mantener propiedades devueltas por funcion useChat que está dentro de mod
  //         onSendMessage: vi.fn(), // Mockea solo onSendMessage
  //       })),
  //     };
  //   });

  //   const wrapper = mount(IndecisionView, {
  //     global: {
  //       stubs: {
  //         ChatMessages: chatMockMessages,
  //       },
  //     },
  //   });
  //   console.log(wrapper.vm); // muestra { onSendMessage: [AsyncFunction: onSendMessage] } solo si desde IndecisionView.vue onSendMessage se expone con defineExpose y cuando no se tiene vi.mock('@/composables/useChat'

  //   // (wrapper.vm as any).onSendMessage = onSendMessageMock; // esto no funciona

  //   // Simular evento personalizado
  //   const messageBoxComponent = wrapper.findComponent(MessageBox); // messageBoxComponent es una instancia de MessageBox
  //   const message = 'Hola Mundo';
  //   messageBoxComponent.vm.$emit('sendMessage', message); // Esto hace que, al emitir el evento sendMessage (send-message del lado de vue y html) desde MessageBox, se ejecute la funcion asociada a dicho evento. Esta funcion sería onSendMessage la cual va a recibir el contenido de message

  //   // await new Promise((r) => setTimeout(r, 150));

  //   expect((wrapper.vm as any).onSendMessage).toHaveBeenCalled();
  //   expect((wrapper.vm as any).onSendMessage).toHaveBeenCalledOnce();
  //   expect((wrapper.vm as any).onSendMessage).toHaveBeenCalledTimes(1);
  //   expect((wrapper.vm as any).onSendMessage).toHaveBeenCalledWith(message);
  // });
});
