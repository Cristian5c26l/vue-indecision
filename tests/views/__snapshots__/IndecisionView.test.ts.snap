// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<IndecisionView /> > should renders chat messages and messagebox correctly 1`] = `
"<div class="bg-gray-100 h-screen flex flex-col max-w-lg mx-auto">
  <div class="bg-blue-500 p-4 text-white flex items-center gap-4"><img class="w-12 h-12 object-cover rounded-full" src="https://media.istockphoto.com/id/1437816897/photo/business-woman-manager-or-human-resources-portrait-for-career-success-company-we-are-hiring.jpg?s=612x612&amp;w=0&amp;k=20&amp;c=tyLvtzutRh22j9GqSGI33Z4HpIwv9vL_MZw_xOE19NQ%3D" alt="">
    <div class="dark:text-white"><span>Mi esposa</span><!-- span or div does work! -->
      <div class="text-sm text-gray-500 dark:text-gray-300">En línea</div>
    </div>
  </div>
  <!--ChatMessages-->
  <div class="flex-1 overflow-y-auto p-4">
    <div class="flex flex-col space-y-2">
      <!-- Messages go here -->
      <!-- <ChatBubble :itsMine="true" :message="'Quieres tomar un café?'" />
      <ChatBubble
        :itsMine="false"
        :message="'no'"
        :image="'https://yesno.wtf/assets/no/22-8806dbccb1edf544723b7f095ff722e8.gif'"
      /> -->
      <!-- <ChatBubble
        v-for="message in messages"
        :key="message.id"
        :its-mine="message.itsMine"
        :message="message.message"
        :image="message.image"
      /> -->
      <!-- :is-last="index === messages.length - 1" -->
      <!--v-if-->
    </div>
  </div>
  <!--MessageBox-->
  <div class="bg-white p-4 flex items-center"><input type="text" placeholder="Type your message..." class="flex-1 border rounded-full px-4 py-2 focus:outline-none"><button class="bg-blue-500 text-white rounded-full p-2 ml-2 hover:bg-blue-600 focus:outline-none"><svg width="20px" height="20px" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg" stroke="#ffffff">
        <g id="SVGRepo_bgCarrier" stroke-width="0"></g>
        <g id="SVGRepo_tracerCarrier" stroke-linecap="round" stroke-linejoin="round"></g>
        <g id="SVGRepo_iconCarrier">
          <path d="M11.5003 12H5.41872M5.24634 12.7972L4.24158 15.7986C3.69128 17.4424 3.41613 18.2643 3.61359 18.7704C3.78506 19.21 4.15335 19.5432 4.6078 19.6701C5.13111 19.8161 5.92151 19.4604 7.50231 18.7491L17.6367 14.1886C19.1797 13.4942 19.9512 13.1471 20.1896 12.6648C20.3968 12.2458 20.3968 11.7541 20.1896 11.3351C19.9512 10.8529 19.1797 10.5057 17.6367 9.81135L7.48483 5.24303C5.90879 4.53382 5.12078 4.17921 4.59799 4.32468C4.14397 4.45101 3.77572 4.78336 3.60365 5.22209C3.40551 5.72728 3.67772 6.54741 4.22215 8.18767L5.24829 11.2793C5.34179 11.561 5.38855 11.7019 5.407 11.8459C5.42338 11.9738 5.42321 12.1032 5.40651 12.231C5.38768 12.375 5.34057 12.5157 5.24634 12.7972Z" stroke="#ffffff" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"></path>
        </g>
      </svg></button></div>
</div>"
`;
